Added support for switch style methods.

sgoodgrove opened the pull request:
First of all, thank you for this project I was about to set about implementing my own type-safe enum pattern, when I found this project and see that it implements everything I was going to and more and in a very refined way.

I'm not entirely happy with the pattern matching switch statements available in C#. Therefore I'm proposing a fluent interface solution to the same task using a .When().Then() fluent interface in place of a regular switch/case statement.

I guess I should have opened an issue for this first, but I think code speaks well for developers. I'm happy if this doesn't get merged, but I think it might be a good discussion point.

Also, if you don't think this is suitable, I think it could be implemented as extension methods and go into a "contrib" namespace if you think it has some legs, but don't want it in the project.

I'm providing this snippet to demonstrate how I imagine it being used:

        [Fact]
        public void WhenMatchesLastParameterActionRuns()
        {
            var three = TestEnum.Three;

            var firstActionRun = false;
            var secondActionRun = false;
            var thirdActionRun = false;

            three
                .When(TestEnum.One).Then(() => firstActionRun = true)
                .When(TestEnum.Two).Then(() => secondActionRun = true)
                .When(TestEnum.One, TestEnum.Two, TestEnum.Three).Then(() => thirdActionRun = true);


            firstActionRun.Should().BeFalse();
            secondActionRun.Should().BeFalse();
            thirdActionRun.Should().BeTrue();
        }

aalmada commented:
I also like the fluent syntax but be aware of the performance implications in this case. Each call to `When()` makes a call to `IEnumerable.Contains()` which enumerates the collection until it finds the item or reachs the end of it. This means the collection will enumerated multiple times, unlike the `switch` clause.

aalmada commented on src/SmartEnum/TValue.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Ardalis.SmartEnum
+{
```
This class is not used. It can be removed.

sgoodgrove replied:
Agreed. Removed.

aalmada commented on src/SmartEnum/Core/DoNotExecute.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace Ardalis.SmartEnum.Core
```
The namespace can be simply `Ardalis.SmartEnum`.  This saves the users to have to add `using` clauses or `Core` for all interface uses.

sgoodgrove replied:
I put these classes into a "Core" namespace as they're not intended to be used by users. So when a user does `using Ardalis.SmartEnum` they don't get these somewhat cryptic implementation details in intellisense.

I'm happy to move them to `Ardalis.SmartEnum` if that's where you'd prefer them.

aalmada replied:
I didn't think of that. That's a good idea to keep `Core`.

aalmada commented on src/SmartEnum/Core/Execute.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace Ardalis.SmartEnum.Core
```
The namespace can be simply `Ardalis.SmartEnum`.  This saves the users to have to add `using` clauses or `Core` for all interface uses.

aalmada commented on src/SmartEnum/Core/IThen.cs:
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Ardalis.SmartEnum.Core
```
The namespace can be simply `Ardalis.SmartEnum`.  This saves the users to have to add `using` clauses or `Core` for all interface uses.

aalmada commented:
The current implementation runs through all the `When()` conditions even when one of them validates `true`. Is it possible to change it to stop looking in this case?

This would also allow a `default` case for when all validate to `false`.

aalmada commented:
The entities `Execute` and `DoNotExecute` are classes so they are allocated on the heap, which results in work for the garbage collector.

Changing them to `struct`, by itself, doesn't fix this as the method `When()` returns an interface which would make the value type to be boxed and copied into the heap.

One possible solution is to delete `DoNotExecute` and `IThen`. Change `Execute` to the following:

```csharp
    public readonly struct Execute<TEnum, TValue>
        where TEnum : SmartEnum<TEnum, TValue>
        where TValue : IEquatable<TValue>, IComparable<TValue>
    {
        private readonly bool enabled;
        private readonly SmartEnum<TEnum, TValue> smartEnum;

        public Execute(bool enabled, SmartEnum<TEnum, TValue> smartEnum)
        {
            this.enabled = enabled;
            this.smartEnum = smartEnum;
        }

        public SmartEnum<TEnum, TValue> Then(Action doThis)
        {
            if (enabled)
                doThis();

            return smartEnum;
        }
    }
```

And change `When()` to the following:

```csharp
        public Execute<TEnum, TValue> When(params SmartEnum<TEnum, TValue>[] smartEnums) =>
            new Execute<TEnum, TValue>(smartEnums.Contains(this), this);

        public Execute<TEnum, TValue> When(IEnumerable<SmartEnum<TEnum, TValue>> smartEnums) =>
            new Execute<TEnum, TValue>(smartEnums.Contains(this), this);
```

sgoodgrove commented:
Yes, I'd not particularly paid attention to performance, but these are all valid concerns. I will restructure the code to work with structs and removing heap allocations.

sgoodgrove commented:
Ok. Let me know what you think of the changes I've made.

I've gotten rid of the interface and introduced a `When` struct and a `Then` struct. These are required to be able to thread through the `stopEvaluating` parameter. The `stopEvaluating` parameter prevents enum comparisons in the `When()` calls when a match has already been made.

I've also added in the `Default()` method.

I've additionally added in `When(SmartEnum<TEnum, TValue> smartEnumWhen)` to avoid the LINQ call to `.Contains()` when it isn't required.

At the moment I've left those two structs inside the ".Core" namespace, so that they doesn't appear in intellisense for users. I could achieve the same thing using an `internal` modifier, though I tend to avoid that. Let me know what you want to do with this and I'll adjust accordingly.

I've also added in a couple of tests around the `Default()` method.

sgoodgrove commented:
Obviously what I really really want is for this to be "exhaustive" as well... I wonder if we could add another method called `Exhaustive()` that throws an exception if there has been no matches, but I don't know how useful that would be at runtime as opposed to compile time... but perhaps it could be seen as an opt-in alternative to `Default()`.

Clearly this is *very* different behaviour to compile time exhaustive checking.

aalmada commented:
Great! I'll have to check that later.

aalmada commented on src/SmartEnum/Core/SmartEnumThen.cs:
```diff
@@ -0,0 +1,33 @@
+        private readonly SmartEnum<TEnum, TValue> smartEnum;
+        private readonly bool stopEvaluating;
+
+        public SmartEnumThen(bool isMatch, bool stopEvaluating, SmartEnum<TEnum, TValue> smartEnum)
```
The constructor can optionally be set to `internal` so that instances can only be created by the methods `When` and `Then`.

aalmada commented on src/SmartEnum/Core/SmartEnumWhen.cs:
```diff
@@ -0,0 +1,58 @@
+        private readonly SmartEnum<TEnum, TValue> smartEnum;
+        private readonly bool stopEvaluating;
+
+        public SmartEnumWhen(bool stopEvaluating, SmartEnum<TEnum, TValue> smartEnum)
```
The constructor can optionally be set to `internal` so that instances can only be created by the methods `When` and `Then`.

aalmada commented on src/SmartEnum/Core/SmartEnumWhen.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <param name="smartEnumWhen">A collection of <see cref="SmartEnum{TEnum, TValue}"/> values to compare to this instance.</param>
+        /// <returns>A executor object to execute a supplied action.</returns>
+        public SmartEnumThen<TEnum, TValue> When(SmartEnum<TEnum, TValue> smartEnumWhen) =>
+            new SmartEnumThen<TEnum, TValue>(isMatch: smartEnumWhen == smartEnum, stopEvaluating: stopEvaluating, smartEnum: smartEnum);
```
Using `smartEnum.Equals(smartEnumWhen )` instead of `smartEnumWhen == smartEnum` saves the `null` check on `smartEnum` performed by the `==` operator.

aalmada commented on src/SmartEnum/SmartEnum.cs:
```diff
@@ -287,27 +289,31 @@ public virtual bool Equals(SmartEnum<TEnum, TValue> other)
+        /// <param name="smartEnumWhen">A collection of <see cref="SmartEnum{TEnum, TValue}"/> values to compare to this instance.</param>
+        /// <returns>A executor object to execute a supplied action.</returns>
+        public SmartEnumThen<TEnum, TValue> When(SmartEnum<TEnum, TValue> smartEnumWhen) =>
+            new SmartEnumThen<TEnum, TValue>(smartEnumWhen == this, false, this);
```
Using `this.Equals(smartEnumWhen )` instead of `smartEnumWhen == this` saves the `null` check on `this` performed by the `==` operator.

sgoodgrove commented:
Yes. It is a shame about the performance.
I've been thinking hard about ways to shortcut the operations or avoid the closure, but I can't think of anything. I tried out a couple of alternatives, but they mostly just made the situation worse.

sgoodgrove commented:
I've udpated with the latest suggestions.

Added When/Then benchmark method.
Change access level of When/Then constructors.
Change comparison from `==` to `.Equals()` to avoid null check.

aalmada commented:
@ardalis IMHO This is a nice feature to add. It needs your seal of approval. 😉

ardalis commented:
Agreed, you two are both awesome, thank you!

ardalis commented:
Can you update the README with some instructions for usage (like what you have in the PR description) as part of this PR?

sgoodgrove commented:
> How about the following? Not sure if you want a perf warning front and centre in the readme.



Because of the limitations of pattern matching SmartEnum also provides a fluent interface to help create clean code:

```csharp
testEnumVar
    .When(TestEnum.One).Then(() => ... )
    .When(TestEnum.Two).Then(() => ... )
    .When(TestEnum.Three).Then(() => ... )
    .Default( ... );
}
```

N.B. For performance critical code the fluent interface carries some overhead that you may wish to avoid. See the available [benchmarks](src/SmartEnum.Benchmarks) code for your use case.

ardalis commented:
Sounds fine; small note with link to benchmarks is perfect.

sgoodgrove commented:
README.md updated
